### ISSUE TITLE
Building for Raspberry Pi

### ISSUE THREAD

**Issue by luRaichu** [hikari-no-yume/touchHLE#258]
I'm attempting to build [this touchHLE fork](https://github.com/alula/touchHLE/tree/star-frog) on my Raspberry Pi Zero 2 W. Here's the log, it completes after ~15 minutes.
```
gay@j:~/touchHLE $ ~/.cargo/bin/cargo run --release
   Compiling serde_derive v1.0.148
   Compiling thiserror-impl v1.0.37
   Compiling proc-macro-error v1.0.4
   Compiling sdl2-sys v0.35.2 (https://github.com/hikari-no-yume/rust-sdl2.git?tag=touchHLE-2#15c58be7)
   Compiling clap_derive v3.2.18
   Compiling thiserror v1.0.37
   Compiling csv-core v0.1.10
   Compiling atty v0.2.14
   Compiling textwrap v0.16.0
   Compiling termcolor v1.2.0
   Compiling itoa v0.4.8
   Compiling time-core v0.1.0
   Compiling strsim v0.10.0
   Compiling once_cell v1.17.0
   Compiling unicode-width v0.1.10
   Compiling bitflags v1.3.2
   Compiling crc32fast v1.3.2
   Compiling clap v3.2.23
   Compiling getopts v0.2.21
   Compiling time-macros v0.2.6
   Compiling serde v1.0.148
   Compiling anyhow v1.0.68
   Compiling libc v0.2.137
   Compiling indexmap v1.9.2
   Compiling byteorder v1.4.3
   Compiling itoa v1.0.4
   Compiling crossbeam-utils v0.8.14
   Compiling ansi_term v0.12.1
   Compiling adler v1.0.2
   Compiling miniz_oxide v0.6.2
   Compiling time v0.3.17
   Compiling cargo-platform v0.1.2
   Compiling camino v1.1.2
   Compiling serde_json v1.0.91
   Compiling bstr v0.2.17
   Compiling semver v1.0.16
   Compiling toml v0.5.11
   Compiling csv v1.1.6
   Compiling cargo_metadata v0.15.3
   Compiling touchHLE_gl_bindings v0.2.1 (/home/pie/touchHLE/src/gles/gl_bindings)
   Compiling cargo-license v0.5.1
   Compiling log v0.4.17
   Compiling touchHLE_dynarmic_wrapper v0.2.1 (/home/pie/touchHLE/src/cpu/dynarmic_wrapper)
   Compiling touchHLE_openal_soft_wrapper v0.2.1 (/home/pie/touchHLE/src/audio/openal_soft_wrapper)
   Compiling touchHLE_dr_mp3_wrapper v0.2.1 (/home/pie/touchHLE/src/audio/dr_mp3_wrapper)
   Compiling touchHLE_stb_image_wrapper v0.2.1 (/home/pie/touchHLE/src/image/stb_image_wrapper)
   Compiling touchHLE_pvrt_decompress_wrapper v0.2.1 (/home/pie/touchHLE/src/image/pvrt_decompress_wrapper)
   Compiling hashbrown v0.12.3
   Compiling sdl2 v0.35.2 (https://github.com/hikari-no-yume/rust-sdl2.git?tag=touchHLE-2#15c58be7)
   Compiling ttf-parser v0.15.2
   Compiling lazy_static v1.4.0
   Compiling safemem v0.3.3
   Compiling line-wrap v0.1.1
   Compiling owned_ttf_parser v0.15.2
   Compiling touchHLE v0.2.1 (/home/pie/touchHLE)
   Compiling flate2 v1.0.25
error: failed to run custom build command for `touchHLE_dynarmic_wrapper v0.2.1 (/home/pie/touchHLE/src/cpu/dynarmic_wrapper)`

Caused by:
  process didn't exit successfully: `/home/pie/touchHLE/target/release/build/touchHLE_dynarmic_wrapper-36fa28a6c7253975/build-script-build` (exit status: 101)
  --- stdout
  CMAKE_TOOLCHAIN_FILE_armv7-unknown-linux-gnueabihf = None
  CMAKE_TOOLCHAIN_FILE_armv7_unknown_linux_gnueabihf = None
  HOST_CMAKE_TOOLCHAIN_FILE = None
  CMAKE_TOOLCHAIN_FILE = None
  CMAKE_GENERATOR_armv7-unknown-linux-gnueabihf = None
  CMAKE_GENERATOR_armv7_unknown_linux_gnueabihf = None
  HOST_CMAKE_GENERATOR = None
  CMAKE_GENERATOR = None
  CMAKE_PREFIX_PATH_armv7-unknown-linux-gnueabihf = None
  CMAKE_PREFIX_PATH_armv7_unknown_linux_gnueabihf = None
  HOST_CMAKE_PREFIX_PATH = None
  CMAKE_PREFIX_PATH = None
  CMAKE_armv7-unknown-linux-gnueabihf = None
  CMAKE_armv7_unknown_linux_gnueabihf = None
  HOST_CMAKE = None
  CMAKE = None
  running: cd "/home/pie/touchHLE/target/release/build/touchHLE_dynarmic_wrapper-036a5d66484890b2/out/build" && CMAKE_PREFIX_PATH="" "cmake" "/home/pie/touchHLE/src/cpu/dynarmic_wrapper/../../../vendor/dynarmic" "-DDYNARMIC_FRONTENDS=A32" "-DDYNARMIC_WARNINGS_AS_ERRORS=OFF" "-DDYNARMIC_TESTS=OFF" "-DDYNARMIC_USE_BUNDLED_EXTERNALS=ON" "-DCMAKE_INSTALL_PREFIX=/home/pie/touchHLE/target/release/build/touchHLE_dynarmic_wrapper-036a5d66484890b2/out" "-DCMAKE_C_FLAGS= -ffunction-sections -fdata-sections -fPIC -march=armv7-a -mfpu=vfpv3-d16" "-DCMAKE_C_COMPILER=/usr/bin/cc" "-DCMAKE_CXX_FLAGS= -ffunction-sections -fdata-sections -fPIC -march=armv7-a -mfpu=vfpv3-d16" "-DCMAKE_CXX_COMPILER=/usr/bin/c++" "-DCMAKE_ASM_FLAGS= -ffunction-sections -fdata-sections -fPIC -march=armv7-a -mfpu=vfpv3-d16" "-DCMAKE_ASM_COMPILER=/usr/bin/cc" "-DCMAKE_BUILD_TYPE=Release"
  -- The C compiler identification is GNU 10.2.1
  -- The CXX compiler identification is GNU 10.2.1
  -- The ASM compiler identification is GNU
  -- Found assembler: /usr/bin/cc
  -- Detecting C compiler ABI info
  -- Detecting C compiler ABI info - done
  -- Check for working C compiler: /usr/bin/cc - skipped
  -- Detecting C compile features
  -- Detecting C compile features - done
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Check for working CXX compiler: /usr/bin/c++ - skipped
  -- Detecting CXX compile features
  -- Detecting CXX compile features - done
  -- Target architecture: arm
  -- Found Boost: /usr/lib/arm-linux-gnueabihf/cmake/Boost-1.74.0/BoostConfig.cmake (found suitable version "1.74.0", minimum required is "1.57")  
  -- Module support is disabled.
  -- Version: 9.0.0
  -- Build type: Release
  -- CXX_STANDARD: 20
  -- Performing Test has_std_20_flag
  -- Performing Test has_std_20_flag - Success
  -- Performing Test has_std_2a_flag
  -- Performing Test has_std_2a_flag - Success
  -- Required features: cxx_variadic_templates
  -- Configuring done
  -- Generating done
  -- Build files have been written to: /home/pie/touchHLE/target/release/build/touchHLE_dynarmic_wrapper-036a5d66484890b2/out/build
  running: cd "/home/pie/touchHLE/target/release/build/touchHLE_dynarmic_wrapper-036a5d66484890b2/out/build" && MAKEFLAGS="-j --jobserver-fds=7,8 --jobserver-auth=7,8" "cmake" "--build" "." "--target" "install" "--config" "Release"
  Scanning dependencies of target fmt
  [  1%] Building CXX object externals/fmt/CMakeFiles/fmt.dir/src/format.cc.o
  [  2%] Building CXX object externals/fmt/CMakeFiles/fmt.dir/src/os.cc.o
  [  3%] Linking CXX static library libfmt.a
  [  3%] Built target fmt
  Scanning dependencies of target mcl
  [  4%] Building CXX object externals/mcl/src/CMakeFiles/mcl.dir/assert.cpp.o
  [  5%] Linking CXX static library libmcl.a
  [  5%] Built target mcl
  Scanning dependencies of target dynarmic
  [  6%] Building CXX object src/dynarmic/CMakeFiles/dynarmic.dir/cmake_pch.hxx.gch
  [  7%] Building CXX object src/dynarmic/CMakeFiles/dynarmic.dir/common/crypto/aes.cpp.o
  [  8%] Building CXX object src/dynarmic/CMakeFiles/dynarmic.dir/common/crypto/crc32.cpp.o
  [  9%] Building CXX object src/dynarmic/CMakeFiles/dynarmic.dir/common/crypto/sm4.cpp.o
  [ 10%] Building CXX object src/dynarmic/CMakeFiles/dynarmic.dir/common/fp/fused.cpp.o
  [ 11%] Building CXX object src/dynarmic/CMakeFiles/dynarmic.dir/common/fp/op/FPCompare.cpp.o
  [ 12%] Building CXX object src/dynarmic/CMakeFiles/dynarmic.dir/common/fp/op/FPConvert.cpp.o
  [ 13%] Building CXX object src/dynarmic/CMakeFiles/dynarmic.dir/common/fp/op/FPMulAdd.cpp.o
  [ 14%] Building CXX object src/dynarmic/CMakeFiles/dynarmic.dir/common/fp/op/FPRecipEstimate.cpp.o
  [ 15%] Building CXX object src/dynarmic/CMakeFiles/dynarmic.dir/common/fp/op/FPRecipExponent.cpp.o

  --- stderr
  In file included from /home/pie/touchHLE/vendor/dynarmic/src/dynarmic/common/fp/op/FPRecipExponent.cpp:8:
  /home/pie/touchHLE/vendor/dynarmic/externals/mcl/src/../include/mcl/bit/bit_field.hpp: In instantiation of ‘constexpr T mcl::bit::mask() [with unsigned int begin_bit = 52; unsigned int end_bit = 62; T = unsigned int]’:
  /home/pie/touchHLE/vendor/dynarmic/externals/mcl/src/../include/mcl/bit/bit_field.hpp:65:48:   required from ‘constexpr T mcl::bit::get_bits(T) [with unsigned int begin_bit = 52; unsigned int end_bit = 62; T = unsigned int]’
  /home/pie/touchHLE/vendor/dynarmic/src/dynarmic/common/fp/op/FPRecipExponent.cpp:24:59:   required from ‘FPT Dynarmic::FP::{anonymous}::DetermineExponentValue(size_t) [with FPT = long long unsigned int; size_t = unsigned int]’
  /home/pie/touchHLE/vendor/dynarmic/src/dynarmic/common/fp/op/FPRecipExponent.cpp:41:53:   required from ‘FPT Dynarmic::FP::FPRecipExponent(FPT, Dynarmic::FP::FPCR, Dynarmic::FP::FPSR&) [with FPT = long long unsigned int]’
  /home/pie/touchHLE/vendor/dynarmic/src/dynarmic/common/fp/op/FPRecipExponent.cpp:57:64:   required from here
  /home/pie/touchHLE/vendor/dynarmic/externals/mcl/src/../include/mcl/bit/bit_field.hpp:44:29: error: static assertion failed: begin_bit must be smaller than size of T
     44 |     static_assert(begin_bit < bitsizeof<T>, "begin_bit must be smaller than size of T");
        |                   ~~~~~~~~~~^~~~~~~~~~~~~~
  /home/pie/touchHLE/vendor/dynarmic/externals/mcl/src/../include/mcl/bit/bit_field.hpp:45:27: error: static assertion failed: end_bit must be smaller than size of T
     45 |     static_assert(end_bit < bitsizeof<T>, "end_bit must be smaller than size of T");
        |                   ~~~~~~~~^~~~~~~~~~~~~~
  /home/pie/touchHLE/vendor/dynarmic/externals/mcl/src/../include/mcl/bit/bit_field.hpp:47:47: warning: left shift count >= width of type [-Wshift-count-overflow]
     47 |     return ones<end_bit - begin_bit + 1, T>() << begin_bit;
        |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
  /home/pie/touchHLE/vendor/dynarmic/externals/mcl/src/../include/mcl/bit/bit_field.hpp: In instantiation of ‘constexpr T mcl::bit::get_bits(T) [with unsigned int begin_bit = 52; unsigned int end_bit = 62; T = unsigned int]’:
  /home/pie/touchHLE/vendor/dynarmic/src/dynarmic/common/fp/op/FPRecipExponent.cpp:24:59:   required from ‘FPT Dynarmic::FP::{anonymous}::DetermineExponentValue(size_t) [with FPT = long long unsigned int; size_t = unsigned int]’
  /home/pie/touchHLE/vendor/dynarmic/src/dynarmic/common/fp/op/FPRecipExponent.cpp:41:53:   required from ‘FPT Dynarmic::FP::FPRecipExponent(FPT, Dynarmic::FP::FPCR, Dynarmic::FP::FPSR&) [with FPT = long long unsigned int]’
  /home/pie/touchHLE/vendor/dynarmic/src/dynarmic/common/fp/op/FPRecipExponent.cpp:57:64:   required from here
  /home/pie/touchHLE/vendor/dynarmic/externals/mcl/src/../include/mcl/bit/bit_field.hpp:65:48:   in ‘constexpr’ expansion of ‘mcl::bit::mask<52, 62, unsigned int>()’
  /home/pie/touchHLE/vendor/dynarmic/externals/mcl/src/../include/mcl/bit/bit_field.hpp:47:47: error: right operand of shift expression ‘(2047 << 52)’ is greater than or equal to the precision 32 of the left operand [-fpermissive]
  /home/pie/touchHLE/vendor/dynarmic/externals/mcl/src/../include/mcl/bit/bit_field.hpp:66:24: warning: right shift count >= width of type [-Wshift-count-overflow]
     66 |     return (value & m) >> begin_bit;
        |            ~~~~~~~~~~~~^~~~~~~~~~~~
  gmake[2]: *** [src/dynarmic/CMakeFiles/dynarmic.dir/build.make:218: src/dynarmic/CMakeFiles/dynarmic.dir/common/fp/op/FPRecipExponent.cpp.o] Error 1
  gmake[1]: *** [CMakeFiles/Makefile2:1047: src/dynarmic/CMakeFiles/dynarmic.dir/all] Error 2
  gmake: *** [Makefile:160: all] Error 2
  thread 'main' panicked at /home/pie/.cargo/registry/src/index.crates.io-1cd66030c949c28d/cmake-0.1.49/src/lib.rs:1104:5:

  command did not execute successfully, got: exit status: 2

  build script failed, must exit now
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
```
I'm currently running the default 32-bit build of Raspbian, but I will try again with a 64-bit image.

**Comment by hikari-no-yume** [comment]
touchHLE only supports 64-bit systems so please try that.

**Comment by luRaichu** [comment]
I can confirm touchHLE builds correctly on 64-bit Raspberry Pi OS, as it's AArch64. Though the build fails unless you also have `libsndio-dev` installed.

**Comment by luRaichu** [comment]
I get a segfault when trying to run an app or open the app picker.
```
touchHLE v0.2.1-110-g42e8b4d (git) — https://touchhle.org/

App bundle info:
- Display name: StarFrog
- Version: 0.118
- Identifier: amaya.starfrog
- Internal name (canonical): starfrog.app
- Minimum OS version: 4.3

Warning: app requires OS version 4.3. Only iPhone OS 2.x and iPhone OS 3.0 apps are currently supported.
Warning: Could not open touchHLE_default_options.txt: No such file or directory (os error 2)
Warning: Could not open touchHLE_options.txt: No such file or directory (os error 2)

touchHLE::environment: Warning: Could not read icon file
touchHLE::gles: Creating an OpenGL ES 1.1 context:
touchHLE::gles: Trying: Native OpenGL ES 1.1
touchHLE::gles: => Success!
touchHLE::window: Driver info: OpenGL ES-CM 1.1 Mesa 20.3.5 / Broadcom / VC4 V3D 2.1
touchHLE::mach_o: Loading armv7 slice for "starfrog"
Unhandled SIGSEGV at pc 0x0000007fa57f59d8
Segmentation fault
```

**Comment by hikari-no-yume** [comment]
Ah, that's unfortunate. I don't have a RPi myself. Maybe gdb or something could be used to get a stack trace for the crash?

**Comment by luRaichu** [comment]
I've never debugged an application before nor used gdb, it scares me...
As for RPi the model I have is only $15 in the US.

**Comment by luRaichu** [comment]
I can't run touchHLE in GDB debugging mode because it segfaults before starting the server. I used this command
```
./touchHLE --gdb=localhost:9001 ~/Games/TouchHLE-Starfrog/starfrog-0118d.ipa
```

**Comment by ciciplusplus** [comment]
Starfrog is not a game which is supported by current version of touchHLE, even of on the trunk. You could check https://appdb.touchhle.org/ for the current compatibility list.

Also, please don't open issues here about problems with unofficial forks.

**Comment by luRaichu** [comment]
The touchHLE fork for running StarFrog isn't radically different from the main version (1 commit ahead, 59 commits behind). Problems there should definitely still apply here.
People have already compiled Alula's touchHLE fork on x86 Windows/Mac and had success there.

**Comment by ciciplusplus** [comment]
Once again, we will not provide support for forks here.

If you can reproduce the issue on the main repo with a supported game, please open another issue, thanks.

**Comment by hikari-no-yume** [comment]
> I can't run touchHLE in GDB debugging mode 

Sorry, to be clear I actually meant running touchHLE in gdb (for debugging touchHLE itself) to get a backtrace, not using touchHLE's gdb mode (for debugging an app running in touchHLE).

**Comment by luRaichu** [comment]
I think I found the problem
```
 Thread 1 "touchHLE" received signal SIGSEGV, Segmentation fault.
0x00000055557f3384 in core::intrinsics::copy_nonoverlapping<u8> (
    src=<optimized out>, dst=<optimized out>, count=<optimized out>)
    at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/intrinsics.rs:2702
2702    /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/intrinsics.rs: No such file or directory.
```

**Comment by hikari-no-yume** [comment]
Can you type `bt` at the gdb prompt to get a backtrace?

**Comment by luRaichu** [comment]
```
#0  0x00000055557f3384 in core::intrinsics::copy_nonoverlapping<u8> (src=<optimized out>, dst=<optimized out>, count=<optimized out>)
    at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/intrinsics.rs:2702
#1  core::ptr::write_unaligned<u32> (dst=<optimized out>, src=<optimized out>)
    at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/ptr/mod.rs:1506
#2  core::ptr::mut_ptr::{impl#0}::write_unaligned<u32> (self=0x0, val=4009754626)
    at /rustc/07dca489ac2d933c78d3c5158e3f43beefeb02ce/library/core/src/ptr/mut_ptr.rs:1547
#3  touchHLE::mem::Mem::write<u32> (self=0x7ffffe9ce8, ptr=..., value=4009754626) at src/mem.rs:454
#4  0x00000055558c16cc in touchHLE::dyld::write_return_to_host_routine (mem=0x7ffffe9ce8, svc=2) at src/dyld.rs:150
#5  0x00000055558c1cf8 in touchHLE::dyld::Dyld::do_initial_linking_with_no_bins (self=0x7ffffea0c0, mem=0x7ffffe9ce8, objc=0x7ffffe9fe0)
    at src/dyld.rs:240
#6  0x0000005555858334 in touchHLE::environment::Environment::new_without_app (options=...) at src/environment.rs:372
#7  0x00000055556292fc in touchHLE::app_picker::show_app_picker_gui (
    options=<error reading variable: Cannot access memory at address 0x70>, apps=...) at src/app_picker.rs:201
#8  0x00000055556286e4 in touchHLE::app_picker::app_picker (options=...) at src/app_picker.rs:74
#9  0x00000055555f786c in touchHLE::main<std::env::Args> (args=...) at src/lib.rs:178

```

**Comment by hikari-no-yume** [comment]
Thanks. I have a suspicion I know what the problem is… could you try running these commands and then seeing if touchHLE will work after that?

* `sudo dd if=/dev/zero of=/swapfile bs=1024 count=4194304`
* `sudo mkswap /swapfile`
* `sudo chmod 600 /swapfile`
* `sudo swapon /swapfile`

This will reserve 4GiB of disk space (on your… SD card, I guess?) and use it as a swap file. touchHLE won't actually _use_ that space, but this might be necessary to get it to start.

If you want to free the space later, do `sudo swapoff /swapfile` and then `sudo rm /swapfile`.

**Comment by luRaichu** [comment]
I already have 2GB of swap but I'll try this

**Comment by luRaichu** [comment]
Okay it runs now on 4GB swap. BUT no window is generated, even when running touchHLE's app picker. and StarFrog crashes
```
touchHLE v0.2.1-110-g42e8b4d (git) — https://touchhle.org/

App bundle info:
- Display name: StarFrog
- Version: 0.118
- Identifier: amaya.starfrog
- Internal name (canonical): starfrog.app
- Minimum OS version: 4.3

Warning: app requires OS version 4.3. Only iPhone OS 2.x and iPhone OS 3.0 apps are currently supported.
No options found for this app in touchHLE_default_options.txt
Using options from touchHLE_options.txt for this app: --preferred-languages=en --landscape-right --button-to-touch=DPadLeft,60,290 --button-to-touch=DPadRight,170,290 --button-to-touch=DPadUp,50,205 --button-to-touch=A,450,290 --button-to-touch=B,380,290 --button-to-touch=Y,420,32 --button-to-touch=X,420,32 --button-to-touch=Start,48,32 --scale-hack=2

touchHLE::environment: Warning: Could not read icon file
touchHLE::gles: Creating an OpenGL ES 1.1 context:
touchHLE::gles: Trying: Native OpenGL ES 1.1
touchHLE::gles: => Success!
touchHLE::window: Driver info: OpenGL ES-CM 1.1 Mesa 20.3.5 / Mesa/X.org / llvmpipe (LLVM 11.0.1, 128 bits)
touchHLE::mach_o: Loading armv7 slice for "starfrog"
touchHLE::mach_o: Bind op: SetDyLibrary(3)
touchHLE::mach_o: Bind op: SetSymbol { name: "_AVAudioSessionCategoryAmbient", flags: (empty) }
touchHLE::mach_o: Bind op: SetSymbolType(Pointer)
touchHLE::mach_o: Bind op: SetSegmentOffset { segment_index: 2, segment_offset: 612 }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d264 -> _AVAudioSessionCategoryAmbient
touchHLE::mach_o: Bind op: SetSymbol { name: "_OBJC_CLASS_$_AVAudioSession", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: 1892 }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d9cc -> _OBJC_CLASS_$_AVAudioSession
touchHLE::mach_o: Bind op: SetDyLibrary(4)
touchHLE::mach_o: Bind op: SetSymbol { name: "_OBJC_CLASS_$_UIDevice", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: 32 }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d9f0 -> _OBJC_CLASS_$_UIDevice
touchHLE::mach_o: Bind op: SetSymbol { name: "_OBJC_CLASS_$_UIImage", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: -28 }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d9d8 -> _OBJC_CLASS_$_UIImage
touchHLE::mach_o: Bind op: SetSymbol { name: "_OBJC_CLASS_$_UIView", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: 72 }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7da24 -> _OBJC_CLASS_$_UIView
touchHLE::mach_o: Bind op: SetSymbol { name: "_OBJC_METACLASS_$_UIView", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: -24 }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7da10 -> _OBJC_METACLASS_$_UIView
touchHLE::mach_o: Bind op: SetDyLibrary(5)
touchHLE::mach_o: Bind op: SetSymbol { name: "_OBJC_CLASS_$_NSAutoreleasePool", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: -88 }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d9bc -> _OBJC_CLASS_$_NSAutoreleasePool
touchHLE::mach_o: Bind op: SetSymbol { name: "_OBJC_CLASS_$_NSBundle", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: 4 }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d9c4 -> _OBJC_CLASS_$_NSBundle
touchHLE::mach_o: Bind op: SetSymbol { name: "_OBJC_CLASS_$_NSFileManager", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: 12 }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d9d4 -> _OBJC_CLASS_$_NSFileManager
touchHLE::mach_o: Bind op: SetSymbol { name: "_OBJC_CLASS_$_NSNumber", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: 16 }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d9e8 -> _OBJC_CLASS_$_NSNumber
touchHLE::mach_o: Bind op: SetSymbol { name: "_OBJC_CLASS_$_NSString", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: -44 }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d9c0 -> _OBJC_CLASS_$_NSString
touchHLE::mach_o: Bind op: SetDyLibrary(7)
touchHLE::mach_o: Bind op: SetSymbol { name: "_OBJC_CLASS_$_CAEAGLLayer", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: 28 }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d9e0 -> _OBJC_CLASS_$_CAEAGLLayer
touchHLE::mach_o: Bind op: SetDyLibrary(8)
touchHLE::mach_o: Bind op: SetSymbol { name: "_OBJC_CLASS_$_EAGLContext", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: 24 }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d9fc -> _OBJC_CLASS_$_EAGLContext
touchHLE::mach_o: Bind op: SetSymbol { name: "_kEAGLColorFormatRGBA8", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: -1932 }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d274 -> _kEAGLColorFormatRGBA8
touchHLE::mach_o: Bind op: SetSymbol { name: "_kEAGLDrawablePropertyColorFormat", flags: (empty) }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d278 -> _kEAGLDrawablePropertyColorFormat
touchHLE::mach_o: Bind op: SetSymbol { name: "_kEAGLDrawablePropertyRetainedBacking", flags: (empty) }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d27c -> _kEAGLDrawablePropertyRetainedBacking
touchHLE::mach_o: Bind op: SetDyLibrary(9)
touchHLE::mach_o: Bind op: SetSymbol { name: "__ZdaPv", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: -560 }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d050 -> __ZdaPv
touchHLE::mach_o: Bind op: SetSymbol { name: "__ZdlPv", flags: (empty) }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d054 -> __ZdlPv
touchHLE::mach_o: Bind op: SetSymbol { name: "__Znam", flags: (empty) }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d058 -> __Znam
touchHLE::mach_o: Bind op: SetSymbol { name: "__Znwm", flags: (empty) }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d05c -> __Znwm
touchHLE::mach_o: Bind op: SetSymbol { name: "___gxx_personality_sj0", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: 504 }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d258 -> ___gxx_personality_sj0
touchHLE::mach_o: Bind op: SetDyLibrary(11)
touchHLE::mach_o: Bind op: SetSymbol { name: "dyld_stub_binder", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: -108 }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d1f0 -> dyld_stub_binder
touchHLE::mach_o: Bind op: SetDyLibrary(12)
touchHLE::mach_o: Bind op: SetSymbol { name: "__objc_empty_cache", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: 2080 }
touchHLE::mach_o: Bind op: BindAndSkipping { times: 5, skip: 16 }
touchHLE::mach_o: Pointer bind: 0x7da14 -> __objc_empty_cache
touchHLE::mach_o: Pointer bind: 0x7da28 -> __objc_empty_cache
touchHLE::mach_o: Pointer bind: 0x7da3c -> __objc_empty_cache
touchHLE::mach_o: Pointer bind: 0x7da50 -> __objc_empty_cache
touchHLE::mach_o: Pointer bind: 0x7da64 -> __objc_empty_cache
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7da78 -> __objc_empty_cache
touchHLE::mach_o: Bind op: SetSymbol { name: "__objc_empty_vtable", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: -100 }
touchHLE::mach_o: Bind op: BindAndSkipping { times: 5, skip: 16 }
touchHLE::mach_o: Pointer bind: 0x7da18 -> __objc_empty_vtable
touchHLE::mach_o: Pointer bind: 0x7da2c -> __objc_empty_vtable
touchHLE::mach_o: Pointer bind: 0x7da40 -> __objc_empty_vtable
touchHLE::mach_o: Pointer bind: 0x7da54 -> __objc_empty_vtable
touchHLE::mach_o: Pointer bind: 0x7da68 -> __objc_empty_vtable
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7da7c -> __objc_empty_vtable
touchHLE::mach_o: Bind op: SetDyLibrary(13)
touchHLE::mach_o: Bind op: SetSymbol { name: "_NSDefaultRunLoopMode", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: -2068 }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d26c -> _NSDefaultRunLoopMode
touchHLE::mach_o: Bind op: SetSymbol { name: "_OBJC_CLASS_$_NSDate", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: 1900 }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d9dc -> _OBJC_CLASS_$_NSDate
touchHLE::mach_o: Bind op: SetSymbol { name: "_OBJC_CLASS_$_NSDictionary", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: 4 }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d9e4 -> _OBJC_CLASS_$_NSDictionary
touchHLE::mach_o: Bind op: SetSymbol { name: "_OBJC_CLASS_$_NSLocale", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: -24 }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d9d0 -> _OBJC_CLASS_$_NSLocale
touchHLE::mach_o: Bind op: SetSymbol { name: "_OBJC_CLASS_$_NSObject", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: 120 }
touchHLE::mach_o: Bind op: BindAndAddAddress { offset: 16 }
touchHLE::mach_o: Pointer bind: 0x7da4c -> _OBJC_CLASS_$_NSObject
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7da60 -> _OBJC_CLASS_$_NSObject
touchHLE::mach_o: Bind op: SetSymbol { name: "_OBJC_CLASS_$_NSRunLoop", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: -112 }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d9f4 -> _OBJC_CLASS_$_NSRunLoop
touchHLE::mach_o: Bind op: SetSymbol { name: "_OBJC_CLASS_$_NSTimer", flags: (empty) }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d9f8 -> _OBJC_CLASS_$_NSTimer
touchHLE::mach_o: Bind op: SetSymbol { name: "_OBJC_CLASS_$_NSURL", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: -52 }
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7d9c8 -> _OBJC_CLASS_$_NSURL
touchHLE::mach_o: Bind op: SetSymbol { name: "_OBJC_METACLASS_$_NSObject", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: 64 }
touchHLE::mach_o: Bind op: BindAndAddAddress { offset: 36 }
touchHLE::mach_o: Pointer bind: 0x7da0c -> _OBJC_METACLASS_$_NSObject
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7da34 -> _OBJC_METACLASS_$_NSObject
touchHLE::mach_o: Bind op: BindAndAddAddress { offset: 52 }
touchHLE::mach_o: Pointer bind: 0x7da38 -> _OBJC_METACLASS_$_NSObject
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7da70 -> _OBJC_METACLASS_$_NSObject
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x7da74 -> _OBJC_METACLASS_$_NSObject
touchHLE::mach_o: Bind op: SetSymbol { name: "___CFConstantStringClassReference", flags: (empty) }
touchHLE::mach_o: Bind op: AddAddress { offset: 72356 }
touchHLE::mach_o: Bind op: BindAndSkipping { times: 5, skip: 12 }
touchHLE::mach_o: Pointer bind: 0x8f51c -> ___CFConstantStringClassReference
touchHLE::mach_o: Pointer bind: 0x8f52c -> ___CFConstantStringClassReference
touchHLE::mach_o: Pointer bind: 0x8f53c -> ___CFConstantStringClassReference
touchHLE::mach_o: Pointer bind: 0x8f54c -> ___CFConstantStringClassReference
touchHLE::mach_o: Pointer bind: 0x8f55c -> ___CFConstantStringClassReference
touchHLE::mach_o: Bind op: Bind
touchHLE::mach_o: Pointer bind: 0x8f56c -> ___CFConstantStringClassReference
touchHLE::mach_o: Bind op: Done
touchHLE::mach_o: Bind op: Done
touchHLE::mach_o: Loading armv6 slice for "libstdc++.6.dylib"
touchHLE::mach_o: Loading armv6 slice for "libgcc_s.1.dylib"
touchHLE::dyld: Write reloc 0x7d9cc target=0x374f6c50 offset=0
touchHLE::dyld: Write reloc 0x7d9f0 target=0x374f6c70 offset=0
touchHLE::dyld: Write reloc 0x7d9d8 target=0x374f6c90 offset=0
touchHLE::dyld: Write reloc 0x7da24 target=0x374f6cd0 offset=0
touchHLE::dyld: Write reloc 0x7da10 target=0x374f6cc0 offset=0
touchHLE::dyld: Write reloc 0x7d9bc target=0x374f6cf0 offset=0
touchHLE::dyld: Write reloc 0x7d9c4 target=0x374f6d10 offset=0
touchHLE::dyld: Write reloc 0x7d9d4 target=0x374f6d30 offset=0
touchHLE::dyld: Write reloc 0x7d9e8 target=0x374f6d70 offset=0
touchHLE::dyld: Write reloc 0x7d9c0 target=0x374f6d90 offset=0
touchHLE::dyld: Write reloc 0x7d9e0 target=0x374f6dd0 offset=0
touchHLE::dyld: Write reloc 0x7d9fc target=0x374f6df0 offset=0
touchHLE::dyld: Write reloc 0x7d258 target=0x374ec82c offset=0
touchHLE::dyld: Write reloc 0x7da14 target=0x0 offset=0
touchHLE::dyld: Write reloc 0x7da28 target=0x0 offset=0
touchHLE::dyld: Write reloc 0x7da3c target=0x0 offset=0
touchHLE::dyld: Write reloc 0x7da50 target=0x0 offset=0
touchHLE::dyld: Write reloc 0x7da64 target=0x0 offset=0
touchHLE::dyld: Write reloc 0x7da78 target=0x0 offset=0
touchHLE::dyld: Write reloc 0x7da18 target=0x0 offset=0
touchHLE::dyld: Write reloc 0x7da2c target=0x0 offset=0
touchHLE::dyld: Write reloc 0x7da40 target=0x0 offset=0
touchHLE::dyld: Write reloc 0x7da54 target=0x0 offset=0
touchHLE::dyld: Write reloc 0x7da68 target=0x0 offset=0
touchHLE::dyld: Write reloc 0x7da7c target=0x0 offset=0
touchHLE::dyld: Write reloc 0x7d9dc target=0x374f6e10 offset=0
touchHLE::dyld: Write reloc 0x7d9e4 target=0x374f6e30 offset=0
touchHLE::dyld: Write reloc 0x7d9d0 target=0x374f6e50 offset=0
touchHLE::dyld: Write reloc 0x7da4c target=0x374f6c30 offset=0
touchHLE::dyld: Write reloc 0x7da60 target=0x374f6c30 offset=0
touchHLE::dyld: Write reloc 0x7d9f4 target=0x374f6e70 offset=0
touchHLE::dyld: Write reloc 0x7d9f8 target=0x374f6e90 offset=0
touchHLE::dyld: Write reloc 0x7d9c8 target=0x374f6eb0 offset=0
touchHLE::dyld: Write reloc 0x7da0c target=0x374f6c20 offset=0
touchHLE::dyld: Write reloc 0x7da34 target=0x374f6c20 offset=0
touchHLE::dyld: Write reloc 0x7da38 target=0x374f6c20 offset=0
touchHLE::dyld: Write reloc 0x7da70 target=0x374f6c20 offset=0
touchHLE::dyld: Write reloc 0x7da74 target=0x374f6c20 offset=0
touchHLE::dyld: Write reloc 0x8f51c target=0x0 offset=0
touchHLE::dyld: Write reloc 0x8f52c target=0x0 offset=0
touchHLE::dyld: Write reloc 0x8f53c target=0x0 offset=0
touchHLE::dyld: Write reloc 0x8f54c target=0x0 offset=0
touchHLE::dyld: Write reloc 0x8f55c target=0x0 offset=0
touchHLE::dyld: Write reloc 0x8f56c target=0x0 offset=0
touchHLE::dyld: Warning: unhandled external relocation "_NSDefaultRunLoopMode" in "starfrog" at 0x7d26c
touchHLE::dyld: Warning: unhandled external relocation "_kEAGLColorFormatRGBA8" in "starfrog" at 0x7d274
touchHLE::dyld: Warning: unhandled external relocation "_kEAGLDrawablePropertyColorFormat" in "starfrog" at 0x7d278
touchHLE::dyld: Warning: unhandled external relocation "_AVAudioSessionCategoryAmbient" in "starfrog" at 0x7d264
touchHLE::dyld: Warning: unhandled external relocation "_kEAGLDrawablePropertyRetainedBacking" in "starfrog" at 0x7d27c
touchHLE::dyld: Warning: unhandled external relocation "dyld_stub_binder" in "starfrog" at 0x7d1f0
touchHLE::dyld: __Znwm_ptr linked? = 0x374ed5f4
CPU emulation begins now.
touchHLE::libc::cxxabi: TODO: __cxa_atexit(GuestFunction(0x374ef884), 0x0, 0x3748a000) (unimplemented)
touchHLE::libc::cxxabi: TODO: __cxa_atexit(GuestFunction(0x374ef8b8), 0x0, 0x3748a000) (unimplemented)
touchHLE::frameworks::uikit::ui_nib: TODO: UIProxyObject replacement for IBFirstResponder, instance 0x374f7da0 left unreplaced
touchHLE::gles: Creating an OpenGL ES 1.1 context:
touchHLE::gles: Trying: Native OpenGL ES 1.1
touchHLE::gles: => Success!
touchHLE::frameworks::opengles::eagl: Driver info: OpenGL ES-CM 1.1 Mesa 20.3.5 / Mesa/X.org / llvmpipe (LLVM 11.0.1, 128 bits)
touchHLE::frameworks::foundation::ns_locale: The app requested your preferred languages. ["en"] will reported based on your --preferred-languages= option.
touchHLE::libc::posix_io: Warning: read(3, 0xfffffd45, 0x1) read only 0x0 bytes
touchHLE::libc::stdio: Tolerating unrecognized fopen() mode flag: 116
touchHLE::libc::posix_io: Warning: read(3, 0xfffffce4, 0x1) read only 0x0 bytes
thread 'main' panicked at src/mem.rs:335:9:
Attempted null-page access at 0x0 (0x1 bytes)
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Register state immediately after panic:
	 R0: 0xfffffe24	 R1: 0x00000000	 R2: 0xfffffed4	 R3: 0x41000000
	 R4: 0x41000000	 R5: 0x41000000	 R6: 0x00000000	 R7: 0xfffffec8
	 R8: 0x374f88f0	 R9: 0xffffffff	R10: 0x00000000	R11: 0x00000000
	R12: 0x375005f0	 SP: 0xfffffe20	 LR: 0x00002a5d	 PC: 0x374f6c14
Attempting to produce stack trace for main thread:
 0. 0x374f6c14 (PC)
 1. 0x2a5d (LR)
 2. 0x13d35
 3. 0x6a93
 4. [host function]
 5. [host function]
 6. [host function]
 7. 0x2228
Next FP (0x0) is outside the stack.
```

**Comment by hikari-no-yume** [comment]
A lack of window is strange, I've never heard of this happening and have no idea how to debug it. Maybe an X11 versus Wayland thing?

**Comment by luRaichu** [comment]
This version of Raspberry Pi OS still uses X11.

**Comment by hikari-no-yume** [comment]
Hmm:

> Driver info: OpenGL ES-CM 1.1 Mesa 20.3.5 / Mesa/X.org / llvmpipe (LLVM 11.0.1, 128 bits)

I doubt this is the actual problem here, but it's weird it would fall back to software rendering. You don't have GPU drivers available?

**Comment by luRaichu** [comment]
Well, it can also use GLES1-on-GL2 but still no window
```
touchHLE::gles: Creating an OpenGL ES 1.1 context:
touchHLE::gles: Trying: OpenGL ES 1.1 via touchHLE GLES1-on-GL2 layer
touchHLE::gles: => Success!
touchHLE::window: Driver info: OpenGL 3.1 Mesa 20.3.5 / Mesa/X.org / llvmpipe (LLVM 11.0.1, 128 bits)
```
I disabled the Raspberry Pi's DRM VC4 V3D driver because it breaks composite overscan settings. I tried to re-enable it but now it breaks HDMI-out.

**Comment by hikari-no-yume** [comment]
Hmm. Very mysterious. Only thing I can suggest is trying a different desktop environment, maybe one that's Wayland-based.

**Comment by luRaichu** [comment]
I tried Wayland and I still don't get a window

**Comment by hikari-no-yume** [comment]
Any difference if you use `--fullscreen`?

**Comment by luRaichu** [comment]
The only difference with fullscreen is that my CPU use rises to 80% instead of 50%.

**Comment by hikari-no-yume** [comment]
Can you try installing SDL2 and OpenAL Soft with your distribution's package manager and then building with `--no-default-features`?

**Comment by luRaichu** [comment]
I was having the exact same lack of window on my x86_64 build of touchHLE and after making a dynamically linked build like you said I have the window back. My Raspberry Pi is still compiling an arm64 build but it should work there as well.

**Comment by hikari-no-yume** [comment]
Ah, then it's probably a bug or incompatibility with the version of SDL2 that touchHLE is statically linked against. With dynamic linking you get a version provided by your distro that might be newer or at least more compatible.

**Comment by luRaichu** [comment]
I have achieved gaming perfection
![IMG_6018](https://github.com/touchHLE/touchHLE/assets/69605426/703bdd4f-dba4-489a-97f9-6dc1dbc8115b)

**Comment by luRaichu** [comment]
Now, I'd like to use my keyboard to control the game, but it seems touchHLE only supports gamepads. on Windows you can use key2joy but I haven't found an equivalent for Linux yet.

**Comment by hikari-no-yume** [comment]
I'm glad you were able to get it working. I'll close this for now; I don't have anything to suggest for the keyboard thing and the original problem is resolved. Hope you'll find some solution. I guess keyboard support could be added to touchHLE but there's quite few games that it would really be suited to.

**Comment by luRaichu** [comment]
> Thanks. I have a suspicion I know what the problem is… could you try running these commands and then seeing if touchHLE will work after that?
> 
>     * `sudo dd if=/dev/zero of=/swapfile bs=1024 count=4194304`
> 
>     * `sudo mkswap /swapfile`
> 
>     * `sudo chmod 600 /swapfile`
> 
>     * `sudo swapon /swapfile`
> 
> 
> This will reserve 4GiB of disk space (on your… SD card, I guess?) and use it as a swap file. touchHLE won't actually _use_ that space, but this might be necessary to get it to start.
> 
> If you want to free the space later, do `sudo swapoff /swapfile` and then `sudo rm /swapfile`.

btw... DO NOT use these commands to increase swap space on your Raspberry Pi. Please use dphys-swapfile instead.

**Comment by luRaichu** [comment]
I tried modifying window.rs to use keyboard input instead of the gamepad, but I'm not a Rust programmer and I failed miserably. What exactly would I need to do to achieve the desired modification? Thanks

**Comment by corped12** [comment]
i would recommend using it with Wine (install it with Pi-Apps)

**Comment by luRaichu** [comment]
> i would recommend using it with Wine (install it with Pi-Apps)

You could, but you'd have reduced performance especially on the lower end Pi's.
